zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/InboundService.kt

```diff
@@ -68,47 +68,88 @@ internal class InboundService<T : ZiplineService>(
+      runCatching {
         function.callSuspending(service, args)
       }
-      endpoint.callCodec.nextOutboundCallCallback = { callbackCall ->
-        endpoint.eventListener.callEnd(
-          externalCall,
-          CallResult(result, callbackCall.encodedCall, callbackCall.serviceNames),
-          callStart,
-        )
-      }
+    }
 
-      // Don't resume a continuation if the Zipline has since been closed.
-      endpoint.scope.ensureActive()
+    if (deferred.isActive) {
+      val cancelCallback = object : CancelCallback, HasPassByReferenceName {
+        override var passbyReferenceName: String? = null
 
-      when {
-        result.isSuccess -> suspendCallback.success(result.getOrNull())
-        else -> suspendCallback.failure(result.exceptionOrNull()!!)
+        override fun cancel() {
+          deferred.cancel()
+        }
+
+        override fun toString() = "CancelCallback/$internalCall"
       }
-    }
 
-    val cancelCallback = object : CancelCallback, HasPassByReferenceName {
-      override var passbyReferenceName: String? = null
+      deferred.invokeOnCompletion {
+        val name = cancelCallback.passbyReferenceName
+        if (name != null) endpoint.remove(name)
+
+        val failure = deferred.getCompletionExceptionOrNull()
+        val kotlinResult = when {
+          failure != null -> Result.failure(failure)
+          else -> deferred.getCompleted()
+        }
+
+        // Don't resume a continuation if the Zipline has since been closed.
+        if (!endpoint.scope.isActive) return@invokeOnCompletion
+
+        endpoint.callCodec.nextOutboundCallCallback = { callbackCall ->
+          endpoint.eventListener.callEnd(
+            externalCall,
+            CallResult(kotlinResult, callbackCall.encodedCall, callbackCall.serviceNames),
+            callStart,
+          )
+        }
+
+        when {
+          kotlinResult.isFailure -> suspendCallback.failure(kotlinResult.exceptionOrNull()!!)
+          else -> suspendCallback.success(kotlinResult.getOrNull())
+        }
+      }
 
-      override fun cancel() {
-        job.cancel()
+      val encodedSuspendingResult = endpoint.callCodec.encodeResult(
+        function,
+        SuspendingResult<Unit>(cancelCallback = cancelCallback)
+      )
+      return encodedSuspendingResult.encodedResult
+
+    } else {
+      val failure = deferred.getCompletionExceptionOrNull()
+      val kotlinResult = when {
+        failure != null -> Result.failure(failure)
+        else -> deferred.getCompleted()
       }
 
-      override fun toString() = "CancelCallback/$internalCall"
-    }
-    job.invokeOnCompletion {
-      val name = cancelCallback.passbyReferenceName
-      if (name != null) endpoint.remove(name)
-    }
+      val suspendingResult = when {
+        kotlinResult.isFailure -> SuspendingResult<Unit>(failure = kotlinResult.exceptionOrNull()!!)
+        else -> SuspendingResult(success = kotlinResult.getOrNull())
+      }
```

Comment:
These two `when`'s need merged or something. You create the `kotlin.Result` and then unpack it just to match on it, force unwrap, and then re-wrap it.

Reply:
Ooooh it’s weirder than that. Added this comment:

```
      // Convert our Deferred<Result<T>> into a Result<T>, which will be failure if _either_ the
      // Deferred failed or if it contains a failure Result.
```

Reply:
(Is this possible??)